### PR TITLE
fix: API, web client URL validation inconsistent

### PR DESCRIPTION
**Pull request by jodiezhuu**
## Description
API and web client URL validation are inconsistent

### Checklist
- [X] Web client identifies this URL as invalid

### Steps to Reproduce
1. Go to URL: [http://127.0.0.1:3000/landscapes/[landscape](http://127.0.0.1:3000/landscapes/%5Blandscape) slug]/upload
2. Go to Share Links tab
3. Add a link with test.c as the URL
4. Click Share Files and Links
5. See API error

Screen Shot 2023-05-12 at 3 48 59 PM

### CODE REVIEW

**Review by paulschreiber — CHANGES_REQUESTED** [review]
@jodiezhuu Let's avoid rewriting the whole URL validation function — it's complicated and we don't want to maintain it.

How about this
- file a bug with yup (and submit a PR if we can figure out their code)
- add a custom validation that checks for single-letter TLDs after we run it through yup's existing validator

**Comment by jodiezhuu** [comment]
> @jodiezhuu Let's avoid rewriting the whole URL validation function — it's complicated and we don't want to maintain it.
> 
> How about this
> 
> * file a bug with yup (and submit a PR if we can figure out their code)
> * add a custom validation that checks for single-letter TLDs after we run it through yup's existing validator

noted and updated

File: src/forms/yup.js [hunk 1]
```diff
@@ -19,6 +19,28 @@ import * as yup from 'yup';
 
 const ARRAY_INDEX_REGEX = /\[([^)]+)\]/;
 
+const hasValidTLD = urlString => {
+  try {
+    const TLD_REGEX = /\.([^.]+)$/;
```
Comment (paulschreiber) [inline src/forms/yup.js:24]: issue: this allows any character. TLDs need to end in two or more letters. Perhaps something like `\.[a-z]{2,}$`?

File: src/sharedData/components/SharedDataUpload/ShareDataLinks.js [hunk 2]
```diff
@@ -54,7 +54,14 @@ import { groupDataEntryUploadsByStatus } from './utils';
 
 export const VALIDATION_SCHEMA = yup
   .object({
-    url: yup.string().trim().ensure().transform(transformURL).url().required(),
+    url: yup
+      .string()
+      .trim()
+      .ensure()
+      .transform(transformURL)
+      .validTld()
```
Comment (paulschreiber) [inline src/sharedData/components/SharedDataUpload/ShareDataLinks.js:62]: issue: please add `validTld()` to the other places we call transformURL (GroupForm, KeyInfoStep, RichTextEditor/index.js)

**Review by paulschreiber — CHANGES_REQUESTED** [review]

File: src/forms/yup.js [hunk 3]
```diff
@@ -19,6 +19,28 @@ import * as yup from 'yup';
...
 
+const hasValidTLD = urlString => {
+  try {
+    const TLD_REGEX = /\.[a-z]{2,}$/i;
+    const url = new URL(urlString);
+    const match = url.hostname.match(TLD_REGEX);
+    if (match) {
+      return true;
+    }
+    return false;
```
Comment (josebui) [inline src/forms/yup.js:30]: Suggestion: I think this should do the same
```suggestion
    return !!match;
```

File: src/forms/yup.js [hunk 4]
```diff
@@ -19,6 +19,28 @@ import * as yup from 'yup';
...
+  } catch (e) {
+    return false;
+  }
+};
+
+yup.addMethod(yup.string, 'validTld', function () {
+  return this.test(
+    'validTld',
+    params => ({ key: 'form.validation_url_invalid', params }),
+    value => (!value ? true : hasValidTLD(value))
```
Comment (josebui) [inline src/forms/yup.js:40]: Question: Why is empty valid?

File: src/forms/yup.js (hunk 3 above)
Comment (paulschreiber) [reply src/forms/yup.js:30]: I recommend `Boolean(match)` instead of `!!match`.

File: src/common/components/RichTextEditor/index.js [hunk 5]
```diff
@@ -104,7 +104,7 @@ const withInlines = editor => {
 
   editor.insertText = text => {
     if (text && isUrl(text)) {
-      wrapLinkPartial(editor)(transformURL(text), text);
+      wrapLinkPartial(editor)(transformURL(text).validTld(), text);
```
Comment (josebui) [inline src/common/components/RichTextEditor/index.js:107]: Question: how is that validTld() function executing?, as far as I can see transformURL() returns a string

Comment (jodiezhuu) [reply src/common/components/RichTextEditor/index.js]: I believe you're right. I thought it was validating the input, but looking at it now, I don't think the validTld() is necessary.

**Review by paulschreiber — APPROVED** [review]

**Pull request merged**